# 02 – Explore Recalls

Understand the structure of NHTSA's recalls file before linking it to complaints.

**Data:** `FLAT_RCL_POST_2010.txt` from [NHTSA Datasets and APIs](https://www.nhtsa.gov/nhtsa-datasets-and-apis)
**Data dictionary:** [RCL.txt](https://static.nhtsa.gov/odi/ffdd/rcl/RCL.txt)

In [1]:
import csv
import pandas as pd

pd.set_option("display.max_colwidth", 200)

In [2]:
RECALLS_PATH = "../data/raw/FLAT_RCL_POST_2010.txt"

RCL_COLUMNS = [
    "record_id", "campno", "maketxt", "modeltxt", "yeartxt", "mfgcampno",
    "compname", "mfgname", "bgman", "endman", "rcltypecd", "potaff", "odate",
    "influenced_by", "mfgtxt", "rcdate", "datea", "rpno", "fmvss",
    "desc_defect", "consequence_defect", "corrective_action", "notes",
    "rcl_cmpt_id", "mfr_comp_name", "mfr_comp_desc", "mfr_comp_ptno",
    "do_not_drive", "park_outside",
]

In [3]:
recalls = pd.read_csv(
    RECALLS_PATH,
    sep="\t",
    header=None,
    names=RCL_COLUMNS,
    encoding="utf-8",
    encoding_errors="replace",
    quoting=csv.QUOTE_NONE
)
recalls.shape

/var/folders/y_/75nvqn5s78748v1dk_k8x5480000gn/T/ipykernel_17659/3974085.py:1: DtypeWarning: Columns (17: rpno, 18: fmvss, 24: mfr_comp_name, 25: mfr_comp_desc, 26: mfr_comp_ptno) have mixed types. Specify dtype option on import or set low_memory=False.
  recalls = pd.read_csv(


(245444, 29)

In [4]:
recalls.head(5)

,record_id,campno,maketxt,modeltxt,yeartxt,mfgcampno,compname,mfgname,bgman,endman,...,desc_defect,consequence_defect,corrective_action,notes,rcl_cmpt_id,mfr_comp_name,mfr_comp_desc,mfr_comp_ptno,do_not_drive,park_outside
0,81716,10V178000,FREIGHTLINER,BUSINESS CLASS M2,2007,FL-577,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,DAIMLER TRUCKS NORTH AMERICA,20050119.0,20100416.0,...,"DTNA IS RECALLING CERTAIN MODEL YEAR 2005 THROUGH 2011 FREIGHTLINER BUSINESS CLASS C2, FCCC B2 CHASSIS, S2 CHASSIS, AND MC CHASSIS MANUFACTURED BEGINNING JANUARY 19, 2005 THROUGH APRIL 16, 2010. ...",THE STOP LAMPS MAY NOT PROPERLY INDICATE THAT THE SERVICE BRAKES ARE APPLIED WHICH MAY LEAD TO A CRASH WITHOUT WARNING.,"DTNA WILL NOTIFY OWNERS AND REPAIRS WILL BE PERFORMED BY DTNA DEALERSHIPS FREE OF CHARGE. THE SAFETY RECALL BEGAN ON MAY 27, 2011. OWNERS MAY CONTACT DTNA AT 1-800-547-0712.","DTNA'S RECALL CAMPAIGN NUMBER IS FL-577.OWNERS MAY ALSO CONTACT THE NATIONAL HIGHWAY TRAFFIC SAFETY ADMINISTRATION'S VEHICLE SAFETY HOTLINE AT 1-888-327-4236 (TTY 1-800-424-9153), OR GO TO <A HREF...",35764000218347000000228,NaN,NaN,NaN,No,No
1,81717,10V559000,KZRV,MONTEGO BAY,2009,NaN,EQUIPMENT,"KZRV, L.P.",NaN,NaN,...,"KZRV IS RECALLING CERTAIN RECREATIONAL VEHICLES EQUIPPED WITH DIMPLEX ELECTRAFLAME, SYMPHONY, OR OPTIFLAME BRANDED ELECTRIC FIREPLACES, STOVES, AND FIREPLACE INSERTS. THE PLUG IN REMOTE CONTROL R...",THIS COULD RESULT IN A FIRE.,"KZRV IS WORKING WITH DIMPLEX AND WILL PROVIDE OWNERS A FREE REPLACEMENT PLUG-IN REMOTE CONTROL KIT. THE SAFETY RECALL BEGAN ON FEBRUARY 17, 2011. OWNERS MAY CONTACT DIMPLEX NORTH AMERICA CUSTOME...","OWNERS MAY ALSO CONTACT THE NATIONAL HIGHWAY TRAFFIC SAFETY ADMINISTRATION'S VEHICLE SAFETY HOTLINE AT 1-888-327-4236 (TTY 1-800-424-9153), OR GO TO <A HREF=HTTP://WWW.SAFERCAR.GOV>HTTP://WWW.SAFE...",38202001002769000000329,NaN,NaN,NaN,No,No
2,81718,10V559000,K-Z,NEW VISION,2005,NaN,EQUIPMENT,"KZRV, L.P.",NaN,NaN,...,"KZRV IS RECALLING CERTAIN RECREATIONAL VEHICLES EQUIPPED WITH DIMPLEX ELECTRAFLAME, SYMPHONY, OR OPTIFLAME BRANDED ELECTRIC FIREPLACES, STOVES, AND FIREPLACE INSERTS. THE PLUG IN REMOTE CONTROL R...",THIS COULD RESULT IN A FIRE.,"KZRV IS WORKING WITH DIMPLEX AND WILL PROVIDE OWNERS A FREE REPLACEMENT PLUG-IN REMOTE CONTROL KIT. THE SAFETY RECALL BEGAN ON FEBRUARY 17, 2011. OWNERS MAY CONTACT DIMPLEX NORTH AMERICA CUSTOME...","OWNERS MAY ALSO CONTACT THE NATIONAL HIGHWAY TRAFFIC SAFETY ADMINISTRATION'S VEHICLE SAFETY HOTLINE AT 1-888-327-4236 (TTY 1-800-424-9153), OR GO TO <A HREF=HTTP://WWW.SAFERCAR.GOV>HTTP://WWW.SAFE...",38202000218720000000329,NaN,NaN,NaN,No,No
3,81719,10V559000,KZRV,MONTEGO BAY,2008,NaN,EQUIPMENT,"KZRV, L.P.",NaN,NaN,...,"KZRV IS RECALLING CERTAIN RECREATIONAL VEHICLES EQUIPPED WITH DIMPLEX ELECTRAFLAME, SYMPHONY, OR OPTIFLAME BRANDED ELECTRIC FIREPLACES, STOVES, AND FIREPLACE INSERTS. THE PLUG IN REMOTE CONTROL R...",THIS COULD RESULT IN A FIRE.,"KZRV IS WORKING WITH DIMPLEX AND WILL PROVIDE OWNERS A FREE REPLACEMENT PLUG-IN REMOTE CONTROL KIT. THE SAFETY RECALL BEGAN ON FEBRUARY 17, 2011. OWNERS MAY CONTACT DIMPLEX NORTH AMERICA CUSTOME...","OWNERS MAY ALSO CONTACT THE NATIONAL HIGHWAY TRAFFIC SAFETY ADMINISTRATION'S VEHICLE SAFETY HOTLINE AT 1-888-327-4236 (TTY 1-800-424-9153), OR GO TO <A HREF=HTTP://WWW.SAFERCAR.GOV>HTTP://WWW.SAFE...",38202001002768000000329,NaN,NaN,NaN,No,No
4,81720,10V559000,KZRV,ESCALADE,2005,NaN,EQUIPMENT,"KZRV, L.P.",NaN,NaN,...,"KZRV IS RECALLING CERTAIN RECREATIONAL VEHICLES EQUIPPED WITH DIMPLEX ELECTRAFLAME, SYMPHONY, OR OPTIFLAME BRANDED ELECTRIC FIREPLACES, STOVES, AND FIREPLACE INSERTS. THE PLUG IN REMOTE CONTROL R...",THIS COULD RESULT IN A FIRE.,"KZRV IS WORKING WITH DIMPLEX AND WILL PROVIDE OWNERS A FREE REPLACEMENT PLUG-IN REMOTE CONTROL KIT. THE SAFETY RECALL BEGAN ON FEBRUARY 17, 2011. OWNERS MAY CONTACT DIMPLEX NORTH AMERICA CUSTOME...","OWNERS MAY ALSO CONTACT THE NATIONAL HIGHWAY TRAFFIC SAFETY ADMINISTRATION'S VEHICLE SAFETY HOTLINE AT 1-888-327-4236 (

In [5]:
recalls[["rcltypecd"]].value_counts()

rcltypecd
V            218710
T             20282
E              5990
C               462
Name: count, dtype: int64

In [6]:
vehicle_recalls = recalls[recalls["rcltypecd"] == "V"]

In [7]:
(vehicle_recalls["yeartxt"] == 9999).sum()

np.int64(382)

In [8]:
vehicle_recalls.loc[vehicle_recalls["yeartxt"] == 9999, ["maketxt", "modeltxt", "compname"]].head(10)

,maketxt,modeltxt,compname
393,WEBASTO,HOLLANDIA 700,EQUIPMENT
394,WEBASTO,HOLLANDIA 600,EQUIPMENT
396,WEBASTO,HOLLANDIA TVS 900,EQUIPMENT
397,WEBASTO,HOLLANDIA TVS 900,EQUIPMENT
398,WEBASTO,HOLLANDIA 600,EQUIPMENT
399,WEBASTO,HOLLANDIA 700,EQUIPMENT
435,WEBASTO,HOLLANDIA TVS 900,EQUIPMENT
436,WEBASTO,HOLLANDIA 600,EQUIPMENT
437,WEBASTO,HOLLANDIA 700,EQUIPMENT
487,WEBASTO,HOLLANDIA 600,EQUIPMENT


In [9]:
vehicle_recalls.loc[vehicle_recalls["yeartxt"] == 9999, "compname"].value_counts().head(10)

compname
VISIBILITY:SUN/MOON ROOF ASSEMBLY                                                    255
EQUIPMENT                                                                             75
SERVICE BRAKES, HYDRAULIC:FOUNDATION COMPONENTS:HOSES, LINES/PIPING, AND FITTINGS     12
SUSPENSION:REAR                                                                       12
SERVICE BRAKES                                                                         9
VISIBILITY                                                                             6
SERVICE BRAKES, HYDRAULIC                                                              5
EXTERIOR LIGHTING                                                                      3
EQUIPMENT:APPLIANCE:AIR CONDITIONER                                                    2
EQUIPMENT:ELECTRICAL                                                                   1
Name: count, dtype: int64

In [10]:
vehicle_recalls = vehicle_recalls[vehicle_recalls["yeartxt"] != 9999]

In [11]:
len(vehicle_recalls)

218328

In [12]:
vehicle_recalls["campno"].duplicated().sum()


np.int64(204815)

In [13]:
vehicle_recalls["campno"].nunique()

13513

In [14]:
example = vehicle_recalls["campno"].iloc[0]
vehicle_recalls.loc[vehicle_recalls["campno"] == example, ["campno", "maketxt", "modeltxt", "yeartxt", "compname", "rcdate"]]

,campno,maketxt,modeltxt,yeartxt,compname,rcdate
0,10V178000,FREIGHTLINER,BUSINESS CLASS M2,2007,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
746,10V178000,FREIGHTLINER,BUSINESS CLASS M2,2009,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
747,10V178000,FREIGHTLINER,BUSINESS CLASS M2,2010,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
748,10V178000,FCCC,B2,2010,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
749,10V178000,FCCC,MC,2009,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
750,10V178000,FCCC,S2,2006,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
751,10V178000,FCCC,MC,2006,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
752,10V178000,FCCC,S2,2011,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
753,10V178000,FREIGHTLINER,BUSINESS CLASS M2,2006,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430
754,10V178000,FREIGHTLINER,BUSINESS CLASS M2,2010,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,20100430


In [15]:
vehicle_recalls["rcdate"] = pd.to_datetime(vehicle_recalls["rcdate"], format="%Y%m%d")
vehicle_recalls["rcdate"].dtype

dtype('<M8[us]')

In [16]:
recalls_dedup["rcdate"].min(), recalls_dedup["rcdate"].max()

NameError: name 'recalls_dedup' is not defined

In [ ]:
vehicle_recalls.loc[[747, 754]]

,record_id,campno,maketxt,modeltxt,yeartxt,mfgcampno,compname,mfgname,bgman,endman,...,desc_defect,consequence_defect,corrective_action,notes,rcl_cmpt_id,mfr_comp_name,mfr_comp_desc,mfr_comp_ptno,do_not_drive,park_outside
747,82463,10V178000,FREIGHTLINER,BUSINESS CLASS M2,2010,FL-577,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,DAIMLER TRUCKS NORTH AMERICA,20050119.0,20100416.0,...,"DTNA IS RECALLING CERTAIN MODEL YEAR 2005 THROUGH 2011 FREIGHTLINER BUSINESS CLASS C2, FCCC B2 CHASSIS, S2 CHASSIS, AND MC CHASSIS MANUFACTURED BEGINNING JANUARY 19, 2005 THROUGH APRIL 16, 2010. ...",THE STOP LAMPS MAY NOT PROPERLY INDICATE THAT THE SERVICE BRAKES ARE APPLIED WHICH MAY LEAD TO A CRASH WITHOUT WARNING.,"DTNA WILL NOTIFY OWNERS AND REPAIRS WILL BE PERFORMED BY DTNA DEALERSHIPS FREE OF CHARGE. THE SAFETY RECALL BEGAN ON MAY 27, 2011. OWNERS MAY CONTACT DTNA AT 1-800-547-0712.","DTNA'S RECALL CAMPAIGN NUMBER IS FL-577.OWNERS MAY ALSO CONTACT THE NATIONAL HIGHWAY TRAFFIC SAFETY ADMINISTRATION'S VEHICLE SAFETY HOTLINE AT 1-888-327-4236 (TTY 1-800-424-9153), OR GO TO <A HREF...",35764000708814000000228,NaN,NaN,NaN,No,No
754,82470,10V178000,FREIGHTLINER,BUSINESS CLASS M2,2010,FL-577,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,DAIMLER TRUCKS NORTH AMERICA,20050119.0,20100416.0,...,"DTNA IS RECALLING CERTAIN MODEL YEAR 2005 THROUGH 2011 FREIGHTLINER BUSINESS CLASS C2, FCCC B2 CHASSIS, S2 CHASSIS, AND MC CHASSIS MANUFACTURED BEGINNING JANUARY 19, 2005 THROUGH APRIL 16, 2010. ...",THE STOP LAMPS MAY NOT PROPERLY INDICATE THAT THE SERVICE BRAKES ARE APPLIED WHICH MAY LEAD TO A CRASH WITHOUT WARNING.,"DTNA WILL NOTIFY OWNERS AND REPAIRS WILL BE PERFORMED BY DTNA DEALERSHIPS FREE OF CHARGE. THE SAFETY RECALL BEGAN ON MAY 27, 2011. OWNERS MAY CONTACT DTNA AT 1-800-547-0712.","DTNA'S RECALL CAMPAIGN NUMBER IS FL-577.OWNERS MAY ALSO CONTACT THE NATIONAL HIGHWAY TRAFFIC SAFETY ADMINISTRATION'S VEHICLE SAFETY HOTLINE AT 1-888-327-4236 (TTY 1-800-424-9153), OR GO TO <A HREF...",35764000708864000000228,NaN,NaN,NaN,No,No


In [ ]:
pair = vehicle_recalls.loc[[747, 754]]
pair.loc[:, pair.nunique() > 1]

,record_id,rcl_cmpt_id
747,82463,35764000708814000000228
754,82470,35764000708864000000228


In [ ]:
key = ["campno", "maketxt", "modeltxt", "yeartxt", "compname"]
vehicle_recalls.duplicated(subset=key).sum()

np.int64(110011)

In [ ]:
recalls_dedup = vehicle_recalls.drop_duplicates(subset=key)
len(vehicle_recalls), len(recalls_dedup)

(218328, 108317)

In [ ]:
vehicle_recalls.loc[[747, 754], ["compname", "mfr_comp_name", "mfr_comp_desc", "mfr_comp_ptno"]]

,compname,mfr_comp_name,mfr_comp_desc,mfr_comp_ptno
747,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,NaN,NaN,NaN
754,EXTERIOR LIGHTING:BRAKE LIGHTS:SWITCH,NaN,NaN,NaN


In [ ]:
vehicle_recalls["influenced_by"].value_counts()

influenced_by
MFR     188943
ODI      27785
OVSC      1600
Name: count, dtype: int64

In [ ]:
recalls_dedup.drop_duplicates("campno")["influenced_by"].value_counts()

influenced_by
MFR     12203
ODI      1203
OVSC      107
Name: count, dtype: int64

## Findings

- File is tab-separated, no header, 29 fields (matches RCL.txt).
- Filtered to vehicle recalls (`rcltypecd == "V"`).
- Excluded 382 rows with unknown model year (9999), mostly aftermarket sunroofs and equipment. They can't be matched to complaints by year.
- One campaign (recall) covers many vehicles: 13,513 campaigns across 218,328 rows.
- 110,011 rows were part-level duplicates (same campaign, vehicle, and component; different part). Deduplicated to 108,317 campaign-vehicle-component rows.
- `rcdate` converted to datetime; recalls range from 2010-01 to 2026-09.
- Component names use the same `:` hierarchy as complaints, a good sign for matching.
- Make names can vary for related manufacturers (e.g. FREIGHTLINER vs FCCC).
- 90% of campaigns are manufacturer-initiated (MFR), 9% NHTSA investigation (ODI), 1% compliance testing (OVSC).
- Recalls go back to 2010, so earlier complaint files are needed for linking.